### ISSUE TITLE
No virtualenv installed on Jenkins nodes

### ISSUE THREAD

**Issue by tisto** [plone/jenkins.plone.org#231]
I set up a Jenkins pipeline job the other day that ran on Node4, because that is the only node that has virtualenv installed. This node is gone and the first three nodes do not seem to have virtualenv installed.

@gforcada is virtualenv installed on the machines?

**Comment by tisto** [comment]
@gforcada I investigated this a bit and found out that we are using self-compiled Python versions which makes it impossible to just install virtualenv by installing python-virtualenv. With Python 2.6 gone can we now go back to using the system Python?

**Comment by gforcada** [comment]
@tisto there is no need for that, we have virtualenv available for py27 on ``/srv/python2.7/bin/virtualenv`` for all python 3 versions (3.6 and 3.7) the binaries are enough, and they can be found on ``/srv/python3.6/bin/python`` and ``/srv/python3.7/bin/python``

If that is not enough, please reopen and let's discuss further.

**Comment by gforcada** [comment]
Oh, regarding self compiled or distributions, @Rotonen provided some patches to optimize the binaries so that tests would run faster, and we don't have to rely on other to update to a newer version of the packages either.

**Comment by tisto** [comment]
@gforcada running a Python version on the CI that differs from what most people are using (system Python) is an anti-pattern of CI IMHO. The CI environment should be identical or at least as close as possible to production systems. On a production system you don't want to use self-compiled Python because you are on your own when it comes to security updates and patches.

**Comment by sunew** [comment]
> The CI environment should be identical or at least as close as possible to production systems. 
I agree 110%

**Comment by tisto** [comment]
@gforcada apart from my comments above. Can we create a symlink in /usr/local/bin for virtualenv to /srv/python2.7/.../bin/python to keep the Jenkinsfile simple and clean?

**Comment by Rotonen** [comment]
We're running explicitly UCS2 Python 2.7 compiled with `--enable-optimizations` in production. This is why it also now tries to bleed through from our CI setup to the Plone CI setup.

It's a ~10% runtime benefit for us, but as the tests can run quicker (and probably quick enough) by just running the layers in parallel, this is not strictly necessary for faster runs on 5.2. I recommended the configure flag as the CI is building its Pythons itself anyway and in that case the gain is IMO worth it.

Since this is turning into an open question on how to do Python on the CI infra, I'd actually recommend using pyenv + pyenv-virtualenv and a buildout / pipeline provided `.python-version` file for managing the Pythons, if not using the ones offered by the base distribution. Or just using the distro provided versions.

Since we're doing pipelines now, can we also do pre-release pipelines, which run the tests in a couple of extra ways to hash out any layer leaks there might be? In that sort of a case we could still run all the tests with both kinds of Pythons in a plethora of permutations before a release (or nightly? weekly? monthly?) and have the day to day builds quicker. We can have nice things.

**Comment by tisto** [comment]
@Rotonen could you elaborate your pyenv + pyenv-virtualenv +.python-version solution. I haven't had time to look into this yet. How do you solve moving the path of the virtualenv between pipeline steps?

**Comment by Rotonen** [comment]
@tisto, if I'm understanding what you mean by 'between pipeline steps', that's baked into how pyenv works. It'll store your Pythons and their environments centrally and it uses the shim resolving mechanism to point to those on the fly, based on your `.python-version`, if the shims are in `$PATH`. It can also be installed for system-wide use in the CI servers / containers and not into userspace.

I am assuming the 'steps' do not hop across nodes, but run on the same node. Also hopping across nodes is only an issue if one does modifications of the virtualenvs at test runtime (which can be a valid way to go, but one should generate randomly named environments for those kinds of systems / deployability / packaging tests). If pyenv is a route to be taken, I recommend managing those environments via Ansible or the container configuration.

One nice CI use case I've done with pyenv is to use detox for running the tests of multiple Python versions in parallel with detox (detox is running on Python 2.7 only for now).

```
$ cat .python-version
2.7.15/envs/detox
3.4.9/envs/package-py34
3.5.6/envs/package-py35
3.6.6/envs/package-py36
3.7.0/envs/package-py37
```

This way you will neatly have the correct `pythonX.Y` and `pipX.Y` in your `$PATH` and the new dependencies get cached in the package specific environments on the first time that node runs the test run with those dependencies.

I usually explicitly init pyenv into the shell session via:

`hash pyenv 2>/dev/null && eval "$(pyenv init -)" && eval "$(pyenv virtualenv-init -)" && export PYTHON_CONFIGURE_OPTS="--enable-optimizations"`

This will nicely fail if pyenv is not in `$PATH` and that can be neatly handled in the script with an early abort.

**Comment by Rotonen** [comment]
> Oh, regarding self compiled or distributions, @Rotonen provided some patches to optimize the binaries so that tests would run faster, and we don't have to rely on other to update to a newer version of the packages either.

To clarify, I've not patched Python itself for the CI runs, but rather provided a patch for the Ansible script to pass a `./configure` flag to use profile guided optimisation.

https://github.com/gforcada/ansible-compile-python/commit/f41a201ead7ddb05fb75283955d2dc343b1e44c3

And I'm unsure as to if it's used on 3.7.

https://github.com/gforcada/ansible-compile-python/commit/51c7a30c6b42f53958d81824c2dd4a1639fd94a0

And AFAIK we currently have no neat way to track which Pythons are compiled which way on the CI infrastructure, so all of them should be recompiled anyway, irrespective of which way people prefer the PGO flag to be.

https://github.com/gforcada/ansible-compile-python/commit/60103ba29997f09719fdcdb7566cb717d9316d34

And we should still probably provide Python 2.4 .. 2.7 and Python 3.4 .. 3.7 at the moment.

**Comment by gforcada** [comment]
Regarding system python vs compiled one: is the exact same python the one used on docker/Fedora/Ubuntu/NixOS etc etc? Everyone is deploying on an ever increasing different amount of platforms that who knows how they got their python there.

The symlink for ease virtualenvs is there, and as a extra bonus, although not added, the ansible role that compiles python can now do so for python 2.4, 2.6, 2.7 and all python 3.1 to 3.7 if anyone needs them.

I'm closing it as the issue was about having virtualenv available, and although it was already, now it is even easier to get to it.

Open another issue if you want to discuss other topics :smile:
